pkg/client/events/replay_client.go

```diff
@@ -0,0 +1,244 @@
+	eventsClient client.EventsQueryClient
+	// eventBytesToTypeDecoder is a function which decodes event subscription
+	// message bytes into the type defined by the EventsReplayClient's generic type
+	// parameter.
+	eventBytesToTypeDecoder func([]byte) (T, error)
+	// latestObsvbls is a replay observable with replay buffer size 1,
+	// which holds the "active latest observable" which is notified when
+	// new events are received by the events query client subscription
+	// created in goPublishEvents. This observable (and the one it emits) closes
+	// when the events bytes observable returns an error and is updated with a
+	// new "active" observable after a new events query subscription is created.
+	latestObsvbls observable.ReplayObservable[U]
+	// latestObsvblsReplayPublishCh is the publish channel for latestObsvbls.
+	// It's used to set latestObsvbls initially and subsequently update it, for
+	// example, when the connection is re-established after erroring.
+	latestObsvblsReplayPublishCh chan<- U
+}
+
+// NewEventsReplayClient creates a new EventsReplayClient from the given
+// dependencies, cometWebsocketURL and subscription query string. It requires a
+// decoder function to be provided which decodes event subscription message
+// bytes into the type defined by the EventsReplayClient's generic type parameter.
+//
+// Required dependencies:
+//   - client.EventsQueryClient
+func NewEventsReplayClient[T any, U observable.ReplayObservable[T]](
+	ctx context.Context,
+	deps depinject.Config,
+	cometWebsocketURL string,
+	queryString string,
+	eventBytesToTypeDecoder func([]byte) (T, error),
+) (client.EventsReplayClient[T, U], error) {
+	// Initialise the mapped client
+	rClient := &replayClient[T, U]{
+		endpointURL:             cometWebsocketURL,
+		queryString:             queryString,
+		eventBytesToTypeDecoder: eventBytesToTypeDecoder,
+	}
+	latestObsvbls,
+		latestObsvblsReplayPublishCh := channel.NewReplayObservable[U](
+		ctx,
+		latestReplayBufferSize,
+	)
+	rClient.latestObsvbls = observable.ReplayObservable[U](latestObsvbls)
+	rClient.latestObsvblsReplayPublishCh = latestObsvblsReplayPublishCh
+
+	// Inject dependencies
+	if err := depinject.Inject(deps, &rClient.eventsClient); err != nil {
+		return nil, err
+	}
+
+	// Concurrently publish blocks to the observable emitted by latestObsvbls.
+	go rClient.goPublishEvents(ctx)
+
+	return rClient, nil
+}
+
+// EventsSequence returns a ReplayObservable, with a replay buffer size of 1,
+// which is notified when new events are received by the events query subscription.
+func (rClient *replayClient[T, R]) EventsSequence(ctx context.Context) R {
+	// Get the latest events observable from the replay observable. We only ever
+	// want the last 1 as any prior latest events observable values are closed.
+	// Directly accessing the zeroth index here is safe because the call to Last
+	// is guaranteed to return a slice with at least 1 element.
```

Comment:
```suggestion
	// Get the active event replay observable from replayObsCache. Only the last element
	// is useful as any prior elements are closed replay observables.
	// Directly accessing the zeroth index here is safe because the call to Last
	// is guaranteed to return a slice with at least 1 element.
```